### ISSUE TITLE
array spread will crash the server process

### ISSUE THREAD

**Issue by pkit** [plv8/plv8js#340]
```
do $$ 
  Object.prototype[Symbol.iterator] = function() {
     return {
       next:() => this
     }
  };
  [...({})]; 
$$ language plv8
```
fixed here: https://github.com/v8/v8/commit/1c48d52bb1ee9bb28e146c60eda08cd4afaa5745

**Comment by JerrySievert** [comment]
oof.  thanks for the report.  better now than later.

I'd rather not cherry-pick the commit if I can help it, so I'll look for a specific version after this commit.

**Comment by pkit** [comment]
Maybe use 7.4.288.13 which is used by node 12.x?
AFAIK only 12.x will not crash due to the one above...

**Comment by JerrySievert** [comment]
thanks.  I'll take a look at that when I get back home.  I'll just need to make sure that it's also one of those magic versions that compiles correctly on windows as well.

**Comment by JerrySievert** [comment]
now I'm remembering the last time I ported plv8 to 7.4.

I was having issues where initialization was going into an infinite loop - I'll see if I can get at least to that state.  it was during one of these:

```
	V8::InitializePlatform(platform);
	V8::Initialize();
```

or the ported versions.

**Comment by JerrySievert** [comment]
I'm hitting a hard crash in `v8::internal::GetRandomMmapAddr()` - not sure if it's from me having done a checkout without an additional `gclient sync` or if there's another issue I'm not yet seeing.

I'll need to wait until I can get somewhere where I'm not on a heavily metered connection to force a full download.

in the mean time, I have a branch started: https://github.com/JerrySievert/plv8/tree/v8_7.4

**Comment by pkit** [comment]
I took the version tag from the changelog. Maybe it's not the real version they use.

**Comment by JerrySievert** [comment]
unfortunately, power went out before I could make any further progress, and it looks like it won't be back until some time tomorrow.  I'll see where it gets tomorrow if I can.

**Comment by JerrySievert** [comment]
7.5 hit today.  I'll take a look at that too.

**Comment by pkit** [comment]
have almost the same problem. actually it crashes when trying to allocate global context first time
it looks like some of their new hacks are not initialized properly: `Isolate::Scope isolate_scope(isolate)` anyone?
probably isolate is not initialized
but just looking into `v8/samples/shell.c` I don't see anything strange, plv8 does exactly the same things

**Comment by pkit** [comment]
Two more things:
1. Actual stable right now is 7.4.288.28
2. I have hit the upgrade error thing, I think I know how to fix it, but not sure if the fix is backportable...

**Comment by JerrySievert** [comment]
yup, that's what I was thinking too: the isolate not being initialized.  and yeah, I looked at shell.cc and d8.cc and traced the same logic - the only change I could see that was conceivable was to platform creation.

I ended up with:

```
	std::unique_ptr<v8::Platform> platform = v8::platform::NewDefaultPlatform(
      8,
			v8::platform::IdleTaskSupport::kEnabled,
      v8::platform::InProcessStackDumping::kEnabled, nullptr);
```

upgrade error thing?

**Comment by pkit** [comment]
> upgrade error thing?

#333
it looks like you cannot call `DefineCustomXxxxxVariable` if older version already defined it
i.e. some sort of check and re-define only if not exists is needed

**Comment by JerrySievert** [comment]
ah, yes, that.  I hadn't spent too much time tracking that down currently.  there are a couple of workarounds and I've been trying to decide which one I deem to the be lesser evil.

**Comment by pkit** [comment]
I think I have some ideas: it looks like v8 uses libc++ in various strategic places when compiling d8 (instead of libstdc++)
I have ended with `is_clang=false use_custom_libcxx=false` but maybe that's not enough

**Comment by pkit** [comment]
something like?
```
	if (!alreadyDefined("plv8.start_proc")) {
        DefineCustomStringVariable("plv8.start_proc",
                                   gettext_noop("PLV8 function to run once when PLV8 is first used."),
                                   NULL,
                                   &plv8_start_proc,
                                   NULL,
                                   PGC_USERSET, 0,
#if PG_VERSION_NUM >= 90100
                                   NULL,
#endif
                                   NULL,
                                   NULL);
	}
```

**Comment by JerrySievert** [comment]
interesting - `use_custom_libcxx=false` is an entirely new one to me.

**Comment by pkit** [comment]
> interesting - use_custom_libcxx=false is an entirely new one to me.

NewPlatform symbol was totally different without it

**Comment by JerrySievert** [comment]
yeah, I noticed that on linux, but it's not different on macOS.

**Comment by JerrySievert** [comment]
on the custom define, I'm going to look to see what postgis is doing to get around this - they're using the same pattern of extension versioning.  core pg doesn't have the issue as all versions are tied together: there's no extension update without updating pg itself.

**Comment by pkit** [comment]
postgis:
```

/* #2382 Before trying to create a user GUC, make sure */
--
/* that the name is not already in use. Why would it be in use? */
/* During an upgrade, a prior copy of the PostGIS library will */
/* already be loaded in memory and the GUC already defined. We */
/* can skip GUC definition in this case, so we just return. */
static const char *guc_name = "postgis.backend";
const char *guc_installed = GetConfigOption(guc_name, TRUE, FALSE);
 
/* Uh oh, this GUC name already exists. Ordinarily we could just go on */
/* our way, but the way the postgis.backend works is by using the "assign" */
/* callback to change which backend is in use by flipping a global variable */
/* over. This saves the overhead of looking up the engine every time, at */
/* the expense of the extra complexity. */
if ( guc_installed )
{
  /* In this narrow case the previously installed GUC is tied to the callback in */
  /* the previously loaded library. Probably this is happening during an */
  /* upgrade, so the old library is where the callback ties to. */
  elog(WARNING, "'%s' is currently set to '%s' and cannot be  changed until you reconnect", guc_name, guc_installed);
  return;
}
 
/* Good, the GUC name is not already in use, so this must be a fresh */
/* and clean new load of the library, and we can define the user GUC */
DefineCustomStringVariable( guc_name, /* name */
............................................
```

**Comment by JerrySievert** [comment]
nice find, thanks!

**Comment by pkit** [comment]
hmm, weird
compiled a fresh v8 in docker using 
`V8_OPTIONS = is_component_build=false v8_static_library=true v8_use_snapshot=true v8_use_external_startup_data=false is_clang=false use_custom_libcxx=false`
but still got weird symbols in d8.o 
```
objdump -t -C build/v8/out.gn/x64.release/obj/d8/d8.o | grep NewDefaultPlatform
0000000000000000         *UND*	0000000000000000 v8::platform::NewDefaultPlatform(int, v8::platform::IdleTaskSupport, v8::platform::InProcessStackDumping, std::__1::unique_ptr<v8::TracingController, std::__1::default_delete<v8::TracingController> >)
```

**Comment by JerrySievert** [comment]
which was the behavior I was getting with g++/linux as well, but not with clang/macos

**Comment by JerrySievert** [comment]
at this point, cherry picking is looking more attractive.

**Comment by pkit** [comment]
oh, god, found it:
```
    Platform *platform = v8::platform::NewDefaultPlatform().release();
    V8::InitializePlatform(platform);
```
works like a charm

**Comment by JerrySievert** [comment]
seriously?  oof.

**Comment by pkit** [comment]
try it :)

**Comment by JerrySievert** [comment]
holy hell.  that's ridiculous. great find!

**Comment by JerrySievert** [comment]
I'll commit this now, and get the `GUC` change in after I take a soldering break.

I'll push up a PR a little after that.

**Comment by pkit** [comment]
but you know what's good? server still dies from code above
albeit not immediately :)

**Comment by JerrySievert** [comment]
oof.

**Comment by pkit** [comment]
```
#
# Fatal javascript OOM in invalid array length
#
```
oh, it's much better, at least we get some exception
can be caught probably

**Comment by JerrySievert** [comment]
also on the `GUC` change, there needs to be an existing variable that has a default for it to check, otherwise it returns `NULL`.

**Comment by JerrySievert** [comment]
```
-> 400 	    V8_IMMEDIATE_CRASH();
```

**Comment by pkit** [comment]
```
ResourceConstraints constraints;
constraints.set_max_semi_space_size_in_kb(....);
constraints.set_max_old_space_size(...);
create_params.constraints = constraints;
```
also probably custom arraybyffer allocator is needed

**Comment by JerrySievert** [comment]
I backed out of using a custom arraybuffer allocator a while back, was hoping to stick with the default one.  you can't trap `V8_IMMEDIATE_CRASH()` though.

**Comment by pkit** [comment]
and finally
`isolate->SetOOMErrorHandler(...)`

**Comment by JerrySievert** [comment]
ah, there's the one, thanks.

**Comment by pkit** [comment]
I think `V8_IMMEDIATE_CRASH()` happens because isolate eats all the memory of v8 itself
if it's constrained traps will fire up

**Comment by JerrySievert** [comment]
this is the error I get on linux:

https://travis-ci.org/JerrySievert/plv8/jobs/534577575#L1126

`v8::platform::NewDefaultPlatform(int, v8::platform::IdleTaskSupport, v8::platform::InProcessStackDumping, std::unique_ptr<v8::TracingController, std::default_delete<v8::TracingController> >)` is missing.  is this what you were getting, and fixed with  `use_custom_libcxx=false`

**Comment by pkit** [comment]
yup, symbols get f up with `std::__1::unique_ptr` instead of `std::unique_ptr`
and fixed by `use_custom_libcxx=false`

**Comment by JerrySievert** [comment]
have you tried running `make installcheck`?

**Comment by JerrySievert** [comment]
I get an abort from most tests.

**Comment by pkit** [comment]
It looks like first call to plv8 succeeds and a second one fails

**Comment by pkit** [comment]
Okay, got it further. It seems like now it only fails when handling js or pg exception

**Comment by pkit** [comment]
Okay, got all tests working.

**Comment by JerrySievert** [comment]
this is merged in 3.0alpha, closing this
